semantics/executable-spec/src/Data/Coders.hs

```diff
@@ -320,56 +421,160 @@ decodE x = decodeCount x 0
 
 decodeCount :: Decode w t -> Int -> Decoder s (Int, t)
 decodeCount (Summands nm f) n = (n + 1,) <$> decodeRecordSum nm (\x -> decodE (f x))
-decodeCount (SumD c) n = pure (n + 1, c)
-decodeCount (RecD c) n = decodeRecordNamed "RecD" (const n) (pure (n, c))
-decodeCount From n = do x <- fromCBOR; pure (n, x)
-decodeCount (D dec) n = do x <- dec; pure (n, x)
-decodeCount (ApplyD c g) n = do
-  (i, f) <- decodeCount c (n + hsize g)
+decodeCount (SumD cn) n = pure (n + 1, cn)
+decodeCount (SparseD cn) n = pure(n+1,cn)
+decodeCount (RecD cn) n = decodeRecordNamed "RecD" (const n) (pure (n, cn))
+decodeCount From n = (n,) <$> fromCBOR
+decodeCount (D dec) n = (n,) <$> dec
+decodeCount (ApplyD cn g) n = do
+  (i, f) <- decodeCount cn (n + hsize g)
   y <- decodeClosed g
   pure (i, f y)
 decodeCount (Invalid k) _ = invalidKey k
 decodeCount (Map f x) n = do (m, y) <- decodeCount x n; pure (m, f y)
-
-decodeClosed :: Decode 'Closed t -> Decoder s t
+decodeCount (DD (Dual _enc dec)) n = (n,) <$> dec
+decodeCount (Emit x) n = pure(n,x)
+decodeCount (u@(SparseKeyed _ _ _ _)) n = (n+1,) <$> decodeClosed u
+decodeCount (Ann x) n = do (m,y) <- decodeCount x n; pure(m,pure y)
+decodeCount (ApplyAnn g x) n = do
+  (i,f) <- decodeCount g (n + hsize x)
+  y <- decodeClosed x
+  pure (i,f <*> y)
+
+decodeClosed :: Decode w t -> Decoder s t
```

Comment:
Given this type, It doesn't seem impossible for `SumD` to appear here. Did you mean to have a constraint?

Reply:
I moved this to be a function defined in a where clause, where the only call site is where the ApplyD  Applys constraint holds, so now that case is truly unreachable. We may try some higher order indexes to see if we
can fix this another way.